contracts/Router.sol

```diff
@@ -0,0 +1,188 @@
+pragma solidity ^0.6.6;
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+
+
+interface IERC20 {
+    function balanceOf(address account) external view returns (uint256);
+    function transfer(address recipient, uint256 amount) external returns (bool);
+    function transferFrom(address sender, address recipient, uint256 amount) external returns (bool);
+    function increaseAllowance(address spender, uint256 addedAmount) public virtual returns (bool);
+}
+
+
+contract Router {
+    using SafeMath for uint256;
+
+    event Pledge(uint256 numCampaign, address addr, uint256 amount);
+    event Unpledge(uint256 numCampaign, address addr, uint256 amount);
+    event Rebase(uint256 numCampaign, treeSold, reserveTokenReceived);
+    event WithdrawToken(address token, address to, uint256 amount);
+    event SetReserveToken(address token);
+
+    address constant private TREE = 0xCE222993A7E4818E0D12BC56376c5a60f92A5783;
+    address constant private RESERVE = 0x390a8Fb3fCFF0bB0fCf1F91c7E36db9c53165d17;
+    address constant private DAI = 0x6B175474E89094C44Da98b954EedeAC495271d0F;
+
+    address public gov;
+    IERC20 public tree = IERC20(TREE);
+    IERC20 public reserveToken = IERC20(DAI);
+
+    struct Campaign {
+        uint256 totalPledged;
+        uint256 numPledgers;
+        uint256 treeSold;
+        mapping (uint256 => address) pledgers;
+        mapping (address => uint256) amountsPledged;
+    }
+
+    uint256 private numCampaigns = 1;
+
+    mapping(uint256 => Campaign) private campaigns;
+ 
+    constructor(address _gov) public {
+        gov = _gov;
+    }
+
+
+    function pledge(uint256 _amount, bool max) external payable {
+        require(!Address.isContract(msg.sender), "Must pledge from EOA");
+        if (max) {_amount = reserveToken.balanceOf(msg.sender);}
+        require(_amount > 0, "Must pledge more than 0.");
+        require(reserveToken.balanceOf(msg.sender) >= _amount, "Cannot pledge more reserveToken than held.");
+        reserveToken.transferFrom(msg.sender, address(this), _amount);
+
+        Campaign storage c = campaigns[numCampaigns];
+        c.totalPledged = c.totalPledged + _amount;
+
+        uint256 pledgerId = getPledgerId(numCampaigns, msg.sender);
+        if (pledgerId == 0) {
+            // user has not pledged before
+            pledgerId = c.numPledgers++;
+            c.pledgers[pledgerId] = msg.sender;
+            c.amountsPledged[msg.sender] = _amount;
+        } else {
+            // user has pledged before, add to their total pledged
+            c.amountsPledged[msg.sender] = c.amountsPledged[msg.sender].add(_amount);
+        }
```

Comment:
You don't need to check the pledgers mapping before adding and you don't need to assign the `_amount`, you can simply add it. If the `msg.sender` isn't in the mapping yet, the value for `c.amountsPledged[msg.sender]` will be `0`.

Reply:
The `pledgers` mapping is checked so that `numPledgers` is increased if they haven't pledged.  That way, we know to pay out the pledger as `c.pledgers[pledgerId]` will now contain the incremented `pledgerId`